File: AhccDesktopViewer/Services/AhccSupabaseClient.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace AhccDesktopViewer.Services;

public sealed class AhccMessageRow
{
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("sender_id")]
    public string? SenderId { get; set; }

    [JsonPropertyName("sender_name")]
    public string SenderName { get; set; } = "";

    [JsonPropertyName("content")]
    public string Content { get; set; } = "";

    [JsonPropertyName("created_at")]
    public string? CreatedAt { get; set; }

    [JsonPropertyName("recipient_name")]
    public string? RecipientName { get; set; }

    [JsonPropertyName("session_id")]
    public string SessionId { get; set; } = "";
}

public sealed class AhccFileRow
{
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("session_id")]
    public string SessionId { get; set; } = "";

    [JsonPropertyName("sender_name")]
    public string SenderName { get; set; } = "";

    [JsonPropertyName("file_name")]
    public string FileName { get; set; } = "";

    [JsonPropertyName("mime")]
    public string Mime { get; set; } = "text/markdown";

    [JsonPropertyName("content")]
    public string Content { get; set; } = "";

    [JsonPropertyName("byte_size")]
    public int ByteSize { get; set; }

    [JsonPropertyName("created_at")]
    public string? CreatedAt { get; set; }
}

public sealed class AhccFileRef
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("mime")]
    public string Mime { get; set; } = "text/markdown";
}

/// <summary>
/// Supabase PostgREST for ahcc_messages + ahcc_files (AHCCDV is Supabase-only).
/// </summary>
public sealed class AhccSupabaseClient : IDisposable
{
    public const string NewSessionContent = "new session";
    public const string FileMarkerPrefix = "[AHCC_FILE]";
    public const string SenderHermes = "Hermes";
    public const int MaxTextFileBytes = 512 * 1024;

    private static readonly Regex FileMarkerRegex = new(
        @"^\[AHCC_FILE\]\s*(\{.*\})\s*$",
        RegexOptions.Singleline | RegexOptions.Compiled);

    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(60) };
    private readonly string _root;
    private readonly string _anonKey;

    public AhccSupabaseClient(string baseUrl, string anonKey)
    {
        _root = (baseUrl ?? "").Trim().TrimEnd('/');
        _anonKey = (anonKey ?? "").Trim();
    }

    public bool IsConfigured =>
        !string.IsNullOrWhiteSpace(_root) && !string.IsNullOrWhiteSpace(_anonKey);

    public async Task<string?> FetchLastSessionIdAsync(CancellationToken ct = default)
    {
        if (!IsConfigured) return null;
        var url =
            $"{_root}/rest/v1/ahcc_messages?select=session_id,created_at,content&order=created_at.desc&limit=1";
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        AddAuth(req);
        using var res = await _http.SendAsync(req, ct).ConfigureAwait(false);
        var body = await res.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        if (!res.IsSuccessStatusCode)
            throw new InvalidOperationException($"Supabase last session HTTP {(int)res.StatusCode}: {Truncate(body)}");

        var rows = JsonSerializer.Deserialize<List<AhccMessageRow>>(body) ?? new();
        var sid = rows.Count > 0 ? rows[0].SessionId : null;
        return string.IsNullOrWhiteSpace(sid) ? null : sid;
    }

    public async Task<List<AhccMessageRow>> FetchSessionMessagesAsync(
        string sessionId,
        CancellationToken ct = default)
    {
        if (!IsConfigured || string.IsNullOrWhiteSpace(sessionId))
            return new List<AhccMessageRow>();

        var url =
            $"{_root}/rest/v1/ahcc_messages?select=*&session_id=eq.{Uri.EscapeDataString(sessionId)}&order=created_at.asc";
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        AddAuth(req);
        using var res = await _http.SendAsync(req, ct).ConfigureAwait(false);
        var body = await res.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        if (!res.IsSuccessStatusCode)
            throw new InvalidOperationException($"Supabase history HTTP {(int)res.StatusCode}: {Truncate(body)}");

        return JsonSerializer.Deserialize<List<AhccMessageRow>>(body) ?? new();
    }

    public async Task<AhccFileRow?> FetchFileAsync(string fileId, CancellationToken ct = default)
    {
        if (!IsConfigured || string.IsNullOrWhiteSpace(fileId)) return null;
        var url = $"{_root}/rest/v1/ahcc_files?select=*&id=eq.{Uri.EscapeDataString(fileId)}&limit=1";
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        AddAuth(req);
        using var res = await _http.SendAsync(req, ct).ConfigureAwait(false);
        var body = await res.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        if (!res.IsSuccessStatusCode)
            throw new InvalidOperationException($"Supabase file HTTP {(int)res.StatusCode}: {Truncate(body)}");
        var rows = JsonSerializer.Deserialize<List<AhccFileRow>>(body) ?? new();
        return rows.Count > 0 ? rows[0] : null;
    }

    public static bool TryParseFileMarker(string content, out AhccFileRef? fileRef)
    {
        fileRef = null;
        if (string.IsNullOrWhiteSpace(content)) return false;
        var m = FileMarkerRegex.Match(content.Trim());
        if (!m.Success) return false;
        try
        {
            fileRef = JsonSerializer.Deserialize<AhccFileRef>(m.Groups[1].Value);
            return fileRef != null && !string.IsNullOrWhiteSpace(fileRef.Id);
        }
        catch
        {
            return false;
        }
    }

    public static string FormatFileMarker(string id, string name, string mime) =>
        FileMarkerPrefix + JsonSerializer.Serialize(new AhccFileRef
        {
            Id = id,
            Name = name,
            Mime = mime,
        });

    private void AddAuth(HttpRequestMessage req)
    {
        req.Headers.TryAddWithoutValidation("apikey", _anonKey);
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _anonKey);
    }

    private static string Truncate(string? s) =>
        string.IsNullOrEmpty(s) ? "" : (s.Length <= 240 ? s : s[..240] + "…");

    public void Dispose() => _http.Dispose();
}

File: AndroidHermesCloudChat/app/src/main/java/com/nous/ahcc/data/supabase/AhccSupabaseClient.kt
package com.nous.ahcc.data.supabase

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.util.UUID

/**
 * Supabase PostgREST client for ahcc_messages + ahcc_files.
 */
class AhccSupabaseClient(
    private val baseUrl: String,
    private val anonKey: String,
) {
    private val root = baseUrl.trim().trimEnd('/')
    private val http = HttpClient(CIO) {
        expectSuccess = false
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true; isLenient = true })
        }
    }

    val isConfigured: Boolean
        get() = root.isNotBlank() && anonKey.isNotBlank()

    suspend fun fetchLastSessionId(): String? {
        if (!isConfigured) return null
        val rows: List<AhccMessageRow> = http.get("$root/rest/v1/ahcc_messages") {
            authHeaders()
            url.parameters.append("select", "session_id,created_at,content")
            url.parameters.append("order", "created_at.desc")
            url.parameters.append("limit", "1")
        }.body()
        return rows.firstOrNull()?.sessionId?.takeIf { it.isNotBlank() }
    }

    suspend fun fetchSessionMessages(sessionId: String): List<AhccMessageRow> {
        if (!isConfigured || sessionId.isBlank()) return emptyList()
        return http.get("$root/rest/v1/ahcc_messages") {
            authHeaders()
            url.parameters.append("select", "*")
            url.parameters.append("session_id", "eq.$sessionId")
            url.parameters.append("order", "created_at.asc")
        }.body()
    }

    /**
     * Resolve Hermes session id: last row in Supabase, or create a new one + marker.
     */
    suspend fun resolveOrCreateSession(
        forceNew: Boolean,
        preferredId: String?,
        senderName: String,
        recipientName: String = "Hermes",
    ): String {
        if (!isConfigured) {
            return preferredId?.takeIf { it.isNotBlank() }
                ?: "ahcc_${UUID.randomUUID().toString().replace("-", "").take(16)}"
        }
        if (!forceNew) {
            fetchLastSessionId()?.let { return it }
            preferredId?.takeIf { it.isNotBlank() }?.let { existing ->
                // Prefer local id only when table is empty вЂ” still mark as new session.
                insertMarker(existing, senderName, recipientName)
                return existing
            }
        }
        val newId = "ahcc_${UUID.randomUUID().toString().replace("-", "").take(16)}"
        insertMarker(newId, senderName, recipientName)
        return newId
    }

    suspend fun insertMarker(sessionId: String, senderName: String, recipientName: String = "Hermes") {
        insert(
            sessionId = sessionId,
            senderName = senderName,
            recipientName = recipientName,
            content = NEW_SESSION_CONTENT,
        )
    }

    suspend fun insert(
        sessionId: String,
        senderName: String,
        content: String,
        recipientName: String? = "Hermes",
        senderId: String? = null,
    ) {
        if (!isConfigured || sessionId.isBlank() || content.isBlank()) return
        val res = http.post("$root/rest/v1/ahcc_messages") {
            authHeaders()
            header("Prefer", "return=minimal")
            contentType(ContentType.Application.Json)
            setBody(
                AhccMessageInsert(
                    senderId = senderId,
                    senderName = senderName,
                    content = content,
                    recipientName = recipientName,
                    sessionId = sessionId,
                )
            )
        }
        if (!res.status.isSuccess()) {
            throw IllegalStateException("Supabase insert HTTP ${res.status.value}: ${res.bodyAsText().take(240)}")
        }
    }

    /**
     * Upload a text/markdown file for peer clients. Returns chat marker `[AHCC_FILE]{...}`.
     */
    suspend fun uploadTextFile(
        sessionId: String,
        senderName: String,
        fileName: String,
        text: String,
        mime: String = "text/markdown",
        recipientName: String? = "Hermes",
    ): String {
        if (!isConfigured) throw IllegalStateException("Supabase is not configured")
        if (sessionId.isBlank()) throw IllegalStateException("session_id is empty")
        val bytes = text.toByteArray(Charsets.UTF_8)
        if (bytes.size > MAX_TEXT_FILE_BYTES) {
            throw IllegalStateException("File too large (max ${MAX_TEXT_FILE_BYTES / 1024} KiB)")
        }
        val fileId = UUID.randomUUID().toString()
        val res = http.post("$root/rest/v1/ahcc_files") {
            authHeaders()
            header("Prefer", "return=minimal")
            contentType(ContentType.Application.Json)
            setBody(
                AhccFileInsert(
                    id = fileId,
                    sessionId = sessionId,
                    senderName = senderName,
                    fileName = fileName,
                    mime = mime,
                    content = text,
                    byteSize = bytes.size,
                )
            )
        }
        if (!res.status.isSuccess()) {
            throw IllegalStateException("Supabase file insert HTTP ${res.status.value}: ${res.bodyAsText().take(240)}")
        }
        val marker = formatFileMarker(fileId, fileName, mime)
        insert(
            sessionId = sessionId,
            senderName = senderName,
            content = marker,
            recipientName = recipientName,
        )
        return marker
    }

    suspend fun fetchFile(fileId: String): AhccFileRow? {
        if (!isConfigured || fileId.isBlank()) return null
        val rows: List<AhccFileRow> = http.get("$root/rest/v1/ahcc_files") {
            authHeaders()
            url.parameters.append("select", "*")
            url.parameters.append("id", "eq.$fileId")
            url.parameters.append("limit", "1")
        }.body()
        return rows.firstOrNull()
    }

    fun close() = http.close()

    private fun io.ktor.client.request.HttpRequestBuilder.authHeaders() {
        header("apikey", anonKey)
        header("Authorization", "Bearer $anonKey")
    }

    companion object {
        const val NEW_SESSION_CONTENT = "new session"
        const val FILE_MARKER_PREFIX = "[AHCC_FILE]"
        const val SENDER_DESKTOP = "WPF User"
        const val SENDER_ANDROID = "AndroidChat"
        const val SENDER_HERMES = "Hermes"
        const val MAX_TEXT_FILE_BYTES = 512 * 1024

        fun formatFileMarker(id: String, name: String, mime: String): String =
            """$FILE_MARKER_PREFIX{"id":"$id","name":${jsonString(name)},"mime":${jsonString(mime)}}"""

        fun parseFileMarker(content: String): Triple<String, String, String>? {
            val trimmed = content.trim()
            if (!trimmed.startsWith(FILE_MARKER_PREFIX)) return null
            val jsonPart = trimmed.removePrefix(FILE_MARKER_PREFIX).trim()
            return runCatching {
                val obj = Json { ignoreUnknownKeys = true }.decodeFromString<AhccFileRef>(jsonPart)
                if (obj.id.isBlank()) null else Triple(obj.id, obj.name, obj.mime)
            }.getOrNull()
        }

        private fun jsonString(s: String): String =
            "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\""
    }
}

@Serializable
data class AhccMessageRow(
    val id: String? = null,
    @SerialName("sender_id") val senderId: String? = null,
    @SerialName("sender_name") val senderName: String = "",
    val content: String = "",
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("recipient_name") val recipientName: String? = null,
    @SerialName("session_id") val sessionId: String = "",
)

@Serializable
data class AhccMessageInsert(
    @SerialName("sender_id") val senderId: String? = null,
    @SerialName("sender_name") val senderName: String,
    val content: String,
    @SerialName("recipient_name") val recipientName: String? = null,
    @SerialName("session_id") val sessionId: String,
)

@Serializable
data class AhccFileInsert(
    val id: String,
    @SerialName("session_id") val sessionId: String,
    @SerialName("sender_name") val senderName: String,
    @SerialName("file_name") val fileName: String,
    val mime: String,
    val content: String,
    @SerialName("byte_size") val byteSize: Int,
)

@Serializable
data class AhccFileRow(
    val id: String? = null,
    @SerialName("session_id") val sessionId: String = "",
    @SerialName("sender_name") val senderName: String = "",
    @SerialName("file_name") val fileName: String = "",
    val mime: String = "text/markdown",
    val content: String = "",
    @SerialName("byte_size") val byteSize: Int = 0,
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
data class AhccFileRef(
    val id: String = "",
    val name: String = "",
    val mime: String = "text/markdown",
)

